File: android/app/src/main/java/ru/cmpas/app/presentation/util/PersonName.kt
package ru.cmpas.app.presentation.util

import java.util.Locale

/**
 * Extracts a display first name from profile values stored in either
 * "Имя Фамилия" or the Russian official "Фамилия Имя Отчество" order.
 *
 * The backend currently exposes a single full-name string, so blindly taking
 * the first token is incorrect. The parser first recognises common given names,
 * then uses patronymic and surname morphology, and finally falls back to the
 * first token for international names.
 */
object PersonName {
    private val locale = Locale("ru")

    private val commonGivenNames = setOf(
        "август", "авдей", "адам", "адриан", "азамат", "аким", "алан", "александр",
        "алексей", "али", "алёна", "алевтина", "алина", "алиса", "алла", "альберт",
        "альбина", "амир", "анастасия", "анатолий", "ангелина", "андрей", "анна",
        "антон", "антонина", "аркадий", "арсен", "арсений", "артём", "артемий",
        "арина", "вадим", "валентин", "валентина", "валерий", "валерия", "варвара",
        "василий", "василиса", "вера", "вероника", "виктор", "виктория", "виталий",
        "влад", "владимир", "владислав", "владлена", "всеволод", "вячеслав", "галина",
        "геннадий", "георгий", "герман", "глеб", "григорий", "давид", "даниил",
        "данил", "данила", "дарья", "демид", "денис", "диана", "дмитрий", "ева",
        "евгений", "евгения", "егор", "екатерина", "елена", "елизавета", "жанна",
        "захар", "злата", "зоя", "иван", "игорь", "илья", "инна", "ирина",
        "карина", "кира", "кирилл", "клавдия", "константин", "ксения", "лариса",
        "лев", "леонид", "лидия", "лилия", "любовь", "людмила", "макар", "максим",
        "марк", "маргарита", "марина", "мария", "матвей", "михаил", "надежда",
        "наталья", "никита", "николай", "нина", "оксана", "олег", "олеся", "ольга",
        "павел", "пётр", "полина", "раиса", "регина", "ренат", "римма", "роберт",
        "родион", "роман", "ростислав", "руслан", "савва", "святослав", "светлана",
        "семён", "сергей", "снежана", "софия", "станислав", "степан", "таисия",
        "тамара", "татьяна", "тимофей", "тимур", "ульяна", "фёдор", "юлия", "юрий",
        "яна", "ярослав", "ярослава",
    )

    private val patronymicEndings = listOf(
        "ович", "евич", "ич", "оглы", "улы", "овна", "евна", "ична", "кызы",
    )

    private val surnameEndings = listOf(
        "ов", "ев", "ёв", "ин", "ын", "ова", "ева", "ёва", "ина", "ына",
        "ский", "цкий", "ской", "ская", "цкая", "ской", "ко", "енко", "ук",
        "юк", "чук", "вич", "ич", "дзе", "швили", "ян", "янц", "оглу",
    )

    fun firstName(raw: String?): String? {
        val tokens = raw
            ?.trim()
            ?.split(Regex("[\\s,;]+"))
            ?.map { it.trim { ch -> !ch.isLetter() && ch != '-' } }
            ?.filter { it.isNotBlank() }
            .orEmpty()

        if (tokens.isEmpty()) return null
        if (tokens.size == 1) return displayCase(tokens.first())

        tokens.firstOrNull { normalise(it) in commonGivenNames }
            ?.let { return displayCase(it) }

        val patronymicIndex = tokens.indexOfFirst(::looksLikePatronymic)
        if (patronymicIndex > 0) return displayCase(tokens[patronymicIndex - 1])

        if (tokens.size == 2) {
            val firstLooksSurname = looksLikeSurname(tokens[0])
            val secondLooksSurname = looksLikeSurname(tokens[1])
            return displayCase(
                when {
                    firstLooksSurname && !secondLooksSurname -> tokens[1]
                    secondLooksSurname && !firstLooksSurname -> tokens[0]
                    else -> tokens[0]
                },
            )
        }

        return displayCase(tokens.first())
    }

    private fun looksLikePatronymic(token: String): Boolean {
        val value = normalise(token)
        return patronymicEndings.any(value::endsWith)
    }

    private fun looksLikeSurname(token: String): Boolean {
        val value = normalise(token)
        return value.length >= 4 && surnameEndings.any(value::endsWith)
    }

    private fun normalise(token: String): String = token.lowercase(locale)

    private fun displayCase(token: String): String = token
        .lowercase(locale)
        .split('-')
        .joinToString("-") { part -> part.replaceFirstChar { it.titlecase(locale) } }
}

File: src/lib/person-name.ts
// Извлекает имя для обращения из значений в формате "Имя Фамилия" или
// официальном "Фамилия Имя Отчество". Слепое взятие первого токена даёт
// фамилию для большинства российских ФИО — сначала проверяем словарь
// распространённых имён, затем морфологию отчества/фамилии, и только потом
// откатываемся на первый токен (для нероссийских имён).
// Портировано из android/.../presentation/util/PersonName.kt — держите в
// синхроне при правках.

const COMMON_GIVEN_NAMES = new Set([
    'август', 'авдей', 'адам', 'адриан', 'азамат', 'аким', 'алан', 'александр',
    'алексей', 'али', 'алёна', 'алевтина', 'алина', 'алиса', 'алла', 'альберт',
    'альбина', 'амир', 'анастасия', 'анатолий', 'ангелина', 'андрей', 'анна',
    'антон', 'антонина', 'аркадий', 'арсен', 'арсений', 'артём', 'артемий',
    'арина', 'вадим', 'валентин', 'валентина', 'валерий', 'валерия', 'варвара',
    'василий', 'василиса', 'вера', 'вероника', 'виктор', 'виктория', 'виталий',
    'влад', 'владимир', 'владислав', 'владлена', 'всеволод', 'вячеслав', 'галина',
    'геннадий', 'георгий', 'герман', 'глеб', 'григорий', 'давид', 'даниил',
    'данил', 'данила', 'дарья', 'демид', 'денис', 'диана', 'дмитрий', 'ева',
    'евгений', 'евгения', 'егор', 'екатерина', 'елена', 'елизавета', 'жанна',
    'захар', 'злата', 'зоя', 'иван', 'игорь', 'илья', 'инна', 'ирина',
    'карина', 'кира', 'кирилл', 'клавдия', 'константин', 'ксения', 'лариса',
    'лев', 'леонид', 'лидия', 'лилия', 'любовь', 'людмила', 'макар', 'максим',
    'марк', 'маргарита', 'марина', 'мария', 'матвей', 'михаил', 'надежда',
    'наталья', 'никита', 'николай', 'нина', 'оксана', 'олег', 'олеся', 'ольга',
    'павел', 'пётр', 'полина', 'раиса', 'регина', 'ренат', 'римма', 'роберт',
    'родион', 'роман', 'ростислав', 'руслан', 'савва', 'святослав', 'светлана',
    'семён', 'сергей', 'снежана', 'софия', 'станислав', 'степан', 'таисия',
    'тамара', 'татьяна', 'тимофей', 'тимур', 'ульяна', 'фёдор', 'юлия', 'юрий',
    'яна', 'ярослав', 'ярослава',
]);

const PATRONYMIC_ENDINGS = ['ович', 'евич', 'ич', 'оглы', 'улы', 'овна', 'евна', 'ична', 'кызы'];

const SURNAME_ENDINGS = [
    'ов', 'ев', 'ёв', 'ин', 'ын', 'ова', 'ева', 'ёва', 'ина', 'ына',
    'ский', 'цкий', 'ской', 'ская', 'цкая', 'ко', 'енко', 'ук',
    'юк', 'чук', 'вич', 'ич', 'дзе', 'швили', 'ян', 'янц', 'оглу',
];

function normalise(token: string): string {
    return token.toLowerCase();
}

function displayCase(token: string): string {
    return token
        .toLowerCase()
        .split('-')
        .map(part => (part.length ? part[0].toUpperCase() + part.slice(1) : part))
        .join('-');
}

function looksLikePatronymic(token: string): boolean {
    const value = normalise(token);
    return PATRONYMIC_ENDINGS.some(ending => value.endsWith(ending));
}

function looksLikeSurname(token: string): boolean {
    const value = normalise(token);
    return value.length >= 4 && SURNAME_ENDINGS.some(ending => value.endsWith(ending));
}

/** Given "Мартынов Илья Сергеевич" (or "Илья Мартынов") returns "Илья". */
export function extractFirstName(raw: string | null | undefined): string | null {
    const tokens = (raw || '')
        .trim()
        .split(/[\s,;]+/)
        .map(t => t.replace(/^[^\p{L}-]+|[^\p{L}-]+$/gu, ''))
        .filter(t => t.length > 0);

    if (tokens.length === 0) return null;
    if (tokens.length === 1) return displayCase(tokens[0]);

    const dictionaryHit = tokens.find(t => COMMON_GIVEN_NAMES.has(normalise(t)));
    if (dictionaryHit) return displayCase(dictionaryHit);

    const patronymicIndex = tokens.findIndex(looksLikePatronymic);
    if (patronymicIndex > 0) return displayCase(tokens[patronymicIndex - 1]);

    if (tokens.length === 2) {
        const firstLooksSurname = looksLikeSurname(tokens[0]);
        const secondLooksSurname = looksLikeSurname(tokens[1]);
        if (firstLooksSurname && !secondLooksSurname) return displayCase(tokens[1]);
        if (secondLooksSurname && !firstLooksSurname) return displayCase(tokens[0]);
        return displayCase(tokens[0]);
    }

    return displayCase(tokens[0]);
}
